src/typechecker/FStar.TypeChecker.Rel.fs

```diff
@@ -654,44 +657,111 @@ let flex_uvar_head t =
     | Tm_uvar (u, _) -> u
     | _ -> failwith "Not a flex-uvar"
 
-let destruct_flex_t t wl : flex_t * worklist =
-    let head, args = U.head_and_args t in
+(* Make sure the uvar at the head of t0 is not affected by a
+ * substitution.
+ *
+ * In the case that it is, first solve it to a new appropriate uvar
+ * without a substitution. This function returns t again, though
+ * it is unchanged (the changes only happen in the UF graph).
+ *
+ * The way we generate the new uvar is find the x's in the domain of ?u
+ * that are not affected by the substitution and make a new variable with
+ * only those binders in the domain. The ones that are affected we will
+ * "hoist" by making the type of ?u' an arrow.
+ *
+ * Example:
+ * If we have (x,y |- ?u : ty)[(x:a) <- x, (y:int) <- 42], we will make
+ * ?u' with x in its binders, abstracted over y:
```

Comment:
This comment is great. But I found one aspect of it slightly misleading. It suggests that you have special treatment for variables that are substituted to themselves, choosing not to abstract over them. However, such identity substitutions are rare. It's more that you simply check if a variable is invariant under the substitution, and if it is then it's not abstracted.

Reply:
Good point. I can make that clearer.